src/Compilers/Extension/CompilerPackage.cs

```diff
@@ -0,0 +1,127 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Shell;
+using System.Reflection;
+using System.IO;
+
+namespace Roslyn.Compilers.Extension
+{
+    [ProvideAutoLoad(UIContextGuids.SolutionExists)]
+    public sealed class CompilerPackage : Package
+    {
+        private const string WriteFileExceptionMessage =
+@"{1}
+
+To reload the Roslyn compiler package, close Visual Studio and any MSBuild processes, then restart Visual Studio.";
+
+        protected override void Initialize()
+        {
+            base.Initialize();
+
+            var packagePath = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location);
+
+            string localRegistryRoot;
+            var reg = (ILocalRegistry2)this.GetService(typeof(SLocalRegistry));
+            reg.GetLocalRegistryRoot(out localRegistryRoot);
+            var registryParts = localRegistryRoot.Split('\\');
+
+            // Is it a valid Hive looks similar to:  
+            //  'Software\Microsoft\VisualStudio\14.0'  'Software\Microsoft\VisualStudio\14.0Roslyn'  'Software\Microsoft\VSWinExpress\14.0'
+            if (registryParts.Length >= 4)
+            {
+                var skuName = registryParts[2];
+                var hiveName = registryParts[3];
+                var roslynHive = string.Format(@"{0}.{1}", registryParts[2], registryParts[3]);
+
+                WriteMSBuildFiles(packagePath, roslynHive);
+
+                try
+                {
+                    Microsoft.Build.Evaluation.ProjectCollection.GlobalProjectCollection.DisableMarkDirty = true;
+                    Microsoft.Build.Evaluation.ProjectCollection.GlobalProjectCollection.SetGlobalProperty("RoslynHive", roslynHive);
+                }
+                finally
+                {
+                    Microsoft.Build.Evaluation.ProjectCollection.GlobalProjectCollection.DisableMarkDirty = false;
+                }
+            }
+        }
+
+        private void WriteMSBuildFiles(string packagePath, string hiveName)
+        {
+            // First we want to ensure any existing Roslyn files are deleted so we don't have old stuff floating
+            // aroud and causing troubles
+            foreach (var file in new DirectoryInfo(GetMSBuildPath()).EnumerateFiles($"*Roslyn*{hiveName}*", SearchOption.AllDirectories))
```

Comment:
Whoever originally wrote the compiler package (@agocke? @cston? @KevinRansom?) did this. I'm just maintaining the cleverness.